lib/mix/lib/mix/tasks/escript.install.ex

```diff
@@ -28,21 +39,150 @@ defmodule Mix.Tasks.Escript.Install do
 
   ## Command line options
 
-    * `--sha512` - checks the escript matches the given sha512 checksum
+    * `--sha512` - checks the escript matches the given sha512 checksum. Only
+      applies to installations via URL or local path.
 
     * `--force` - forces installation without a shell prompt; primarily
       intended for automation in build systems like make
 
+    * `--submodules` - fetch repository submodules before building escript from
+      git or github
+
   """
 
   @behaviour Mix.Local.Installer
 
   @escript_file_mode 0o555 # only read and execute permissions
 
-  @switches [force: :boolean, sha512: :string]
+  @switches [force: :boolean, sha512: :string, submodules: :boolean]
   @spec run(OptionParser.argv) :: boolean
   def run(argv) do
-    Mix.Local.Installer.install({__MODULE__, :escript}, argv, @switches)
+    {opts, args} = OptionParser.parse!(argv, strict: @switches)
+
+    case args do
+      ["git" | rest] when rest != [] ->
+        raise_if_sha512("git", opts)
+        install_from_git(rest, opts)
+      ["github" | rest] when rest != [] ->
+        raise_if_sha512("github", opts)
+        install_from_github(rest, opts)
+      ["hex" | rest] when rest != [] ->
+        raise_if_sha512("hex", opts)
+        install_from_hex(rest, opts)
+      _ ->
+        Mix.Local.Installer.install({__MODULE__, :escript}, argv, @switches)
+    end
+  end
+
+  defp raise_if_sha512(name, opts) do
+    if opts[:sha512] do
+      Mix.raise "--sha512 is not supported for escript.install from #{name}"
+    end
+  end
+
+  defp install_from_github([repo | rest], opts) do
+    url = "https://github.com/#{repo}.git"
+    install_from_git([url | rest], opts)
+  end
+
+  defp install_from_git([url], opts) do
+    install_from_git([url, "branch", "master"], opts)
+  end
+
+  defp install_from_git([url, ref_type, ref], opts) do
+    git_opts =
+      ref_to_config(ref_type, ref) ++
+      [git: url, submodules: opts[:submodules]]
+
+    fetch_build_and_install_escript("new escript", {:"new escript", git_opts}, opts)
+  end
+
+  defp install_from_git([_url | rest], _opts) do
+    Mix.raise "escript.install received invalid git checkout spec: #{Enum.join(rest, " ")}"
+  end
+
+  defp ref_to_config("branch", branch), do: [branch: branch]
+
+  defp ref_to_config("tag", tag), do: [tag: tag]
+
+  defp ref_to_config("ref", ref), do: [ref: ref]
+
+  defp ref_to_config(ref_type, _) do
+    Mix.raise "escript.install expected one of \"branch\", \"tag\", or \"ref\". Got: \"#{ref_type}\""
+  end
+
+  defp install_from_hex([package_name], opts) do
+    install_from_hex([package_name, ">= 0.0.0"], opts)
+  end
+
+  defp install_from_hex([package_name, version], opts) do
+    fetch_build_and_install_escript(package_name, {String.to_atom(package_name), version}, opts)
```

Comment:
@ericmj @michalmuskala please do correct me if I'm wrong, but I believe this is unnecessary, since the given package is the only dependency. My reasoning is that because the given package's app name is unused in the "installer" wrapper package, and any dependencies of the given package cannot be dependent on the given package itself, there is no need to specify the actual app name because there can be no packages that actually refer to it.